python/ray/data/tests/test_execution_optimizer.py

```diff
@@ -1698,6 +1699,37 @@ def test_zero_copy_fusion_eliminate_build_output_blocks(ray_start_regular_shared
     )
 
 
+def test_logical_optimization_e2e(ray_start_regular_shared):
+    ds = ray.data.range(10, override_num_blocks=2)
+    ds = ds.randomize_block_order()
+    ds = ds.map_batches(lambda x: x)
+    ds = ds.randomize_block_order()
+    assert set(extract_values("id", ds.take_all())) == set(range(10))
+
+    # Expected execution plan:
+    expected_execution_plan = (
+        "InputDataBuffer[Input] -> "
+        "TaskPoolMapOperator[ReadRange->MapBatches(<lambda>)] -> "
+        "AllToAllOperator[RandomizeBlockOrder]"
+    )
+    assert str(get_execution_plan(ds._plan._logical_plan).dag) == str(
+        expected_execution_plan
+    )
+
+    # check that randomize_block_order is pushed to the end
+    name = "ReadRange->MapBatches(<lambda>)"
+    assert name in ds.stats()
+
+    # check that Operator 2 is the randomize_block_order operator
+    assert "Operator 2 RandomizeBlockOrder:" in ds.stats()
+
+    # check that randomize_block_order is correctly deduped
+    deduped_third_op = "Operator 3 RandomizeBlockOrder:"
+    assert deduped_third_op not in ds.stats()
+
+    _check_usage_record(["ReadRange", "MapBatches", "RandomizeBlockOrder"])
```

Comment:
Let's assert whole stats then, rather than collecting this jigsaw puzzle that is impossible to verify (for reviewer)